ee/agent/storage/sqlite/keyvalue_store_sqlite.go

```diff
@@ -0,0 +1,219 @@
+package agentsqlite
+
+import (
+	"context"
+	"database/sql"
+	"embed"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang-migrate/migrate/v4"
+	sqlitemigrationdriver "github.com/golang-migrate/migrate/v4/database/sqlite"
+	"github.com/golang-migrate/migrate/v4/source/iofs"
+	_ "modernc.org/sqlite"
+)
+
+//go:embed migrations/*.sqlite
+var migrations embed.FS
+
+type SqliteStore struct {
+	conn          *sql.DB
+	rootDirectory string
+}
+
+func NewStore(ctx context.Context, rootDirectory string) (*SqliteStore, error) {
```

Comment:
I think this should take a `tableName` parameter.

Reply:
I'm on the fence about this. I think taking a `tableName` parameter makes this store more flexible. On the other hand, it requires consumers to know info that should be abstracted from them, and it means that it's now possible for a consumer to request a store for a table that doesn't exist.

Reply:
So something like this would address the drawbacks above:

```
const (
	TableKeyValuePairs = "keyvalue_pairs"
)

var supportedTables = map[string]struct{}{
	TableKeyValuePairs: struct{}{},
}

...


func NewStore(ctx context.Context, rootDirectory string, tableName string) (*SqliteStore, error) {
	if _, ok := supportedTables[tableName]; !ok {
		return nil, fmt.Errorf("unsupported table %s", tableName)
	}
...
```

...but now we've either got to standardize the two column names for all possible tables, or take and validate those as arguments too. Feels maybe like overoptimizing at this point?

Reply:
I'll grant that this might be premature optimization, but here's how I structure this in my head.
- We want to store information
- For historical reasons, our information is stored in KV stores, broken up by area/type etc
- (This is implemented in bolt, using buckets, which provide kv stores)
- We don't love bbolt. The store interface is a newish layer on top of bbolt to allow us to play with swapping out the underlying implementation
- We now need to content with multi-accessor
- And also not-private data
- To me, this fits with our existing model -- segment that into "buckets" and treat them like KV stores.
- Yes, that's not as powerful as using all of what SQL can provide, but there's value in sticking with our interfaces
- Which leads me to a place of wanting
   + underlying backing stores (like bbolt and sqlite)
   + consistent store layer build on ^^^
   + deciding which "bucket" uses which backing store. 
  
I see the `ee/agent/storage/sqlite` as being that backing store. But I think it needs that bucket layer on top of it. 

The startup settings stuff feels a natural "bucket" to me.

In terms of signatures, I think `map[string]struct` would be weird, but I can imagine `type supportedStores string` and an list of known ones. I think it's correct to use the same column names -- any KV using sqlite should have the same schema.

Reply:
This makes sense to me, and I'll add it!

> In terms of signatures, I think `map[string]struct` would be weird, but I can imagine `type supportedStores string` and an list of known ones.

I like a map over a list for ease of checking whether a table is supported or not. `map[string]struct{}` is a little better than `map[string]bool` IMO because it's clearer that you're just checking for presence in the map, instead of maybe checking for presence in the map but maybe also needing to check if the bool is true or false.

Reply:
I've updated this!